### PR TITLE
feat: add coupon system with universal/product/user scopes

### PR DESCRIPTION
**Pull request by Yourdaylight**
## Summary

Add native coupon/discount support to Casdoor's SaaS management module with three coupon modes:

- **Universal** — applies to all products and users
- **Product-specific** — restricted to selected products
- **User-specific** — restricted to selected users

Supports both **percentage** (with optional max discount cap) and **fixed amount** discount types, with configurable usage limits, validity period, minimum order amount, and currency matching.

### Backend (Go)
- New `Coupon` and `CouponUsage` models with full CRUD operations
- `ValidateCoupon`: validates state, expiry, stock, scope matching, currency, per-user usage limit
- `CalculateDiscount`: handles percentage (with cap) and fixed amount calculation
- `ApplyCoupon`: atomic usage recording with concurrent safety (uses DB-level `WHERE used_count < quantity`)
- Integrated into `PlaceOrder` flow — coupon discount applied before order price is finalized
- Coupon usage recorded on successful payment (both Balance and external providers via `NotifyPayment`)
- 6 new API endpoints: CRUD + `/api/validate-coupon` for real-time code validation

### Frontend (React)
- `CouponListPage`: sortable/searchable table with scope tags, usage stats, discount display
- `CouponEditPage`: dynamic form — scope-dependent fields (Products/Users selectors), currency with flags, date pickers
- Sidebar menu item under **Business > Coupons**
- i18n support (English + Chinese)

### Tests - 29 test cases
- **CalculateDiscount** (8 cases): percentage/fixed/cap/boundary conditions
- **ValidateCoupon** (18 cases): expired, inactive, out-of-stock, scope matching, min order, currency mismatch, per-user limit
- **ApplyCoupon** (2 cases): usage recording and amount tracking
- **CRUD** (7 cases): add/get/getByCode/update/delete/pagination/code uniqueness

### Coupon List Page
Coupon List

### Coupon Edit Page - Universal Scope
Edit Universal

### Coupon Edit Page - Product Specific Scope
Edit Product

### Coupon Edit Page - User Specific Scope
Edit User

## Test plan
- [x] `go build ./...` compiles successfully
- [x] All 29 unit/integration tests pass
- [x] Frontend compiles and hot-reloads without errors
- [x] Coupon CRUD works via UI (create, edit, delete)
- [x] Scope switching dynamically shows/hides Products/Users fields
- [x] Currency selector uses flag dropdown consistent with Product page
- [x] End-to-end: create coupon > buy product with coupon code > verify discounted price

### CODE REVIEW

**Comment by hsluoyz** [comment]
@Yourdaylight fix linter:

image

File: object/payment.go [hunk 1]
```diff
@@ -418,6 +418,14 @@ func NotifyPayment(body []byte, owner string, paymentName string, lang string) (
...
 			return nil, err
 		}
+
+		// Record coupon usage after successful external payment
+		if order.CouponName != "" {
+			err = ApplyCoupon(order.Owner, order.CouponName, order.User, order.Name, order.CouponDiscount)
+			if err != nil {
+				return nil, err
+			}
+		}
```
[the remark this answers is not in the document]
Comment (Yourdaylight) [reply object/payment.go:428]: This follows the same pattern as UpdateProductStock() , which is also called after 
  payment/order state has been persisted. If stock update fails, the payment is already in a       
  terminal state and webhook retries won't re-run it either. Making coupon usage idempotent would 
  require refactoring the entire NotifyPayment post-payment flow, which I think is better suited
  for a separate PR.

File: object/order_pay.go [hunk 2]
```diff
@@ -369,6 +387,14 @@ func PayOrder(providerName, host, paymentEnv string, order *Order, lang string)
...
 			return nil, nil, err
 		}
+
+		// Record coupon usage after successful balance payment
+		if order.CouponName != "" {
+			err = ApplyCoupon(order.Owner, order.CouponName, order.User, order.Name, order.CouponDiscount)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
```
[the remark this answers is not in the document]
Comment (Yourdaylight) [reply object/order_pay.go:397]: Same as the comment above
